network/topology/cache_test.go

```diff
@@ -52,45 +47,89 @@ func TestCache_GenerateFanout_HappyPath(t *testing.T) {
 	require.Nil(t, fanout)
-	require.Equal(t, cache.fingerprint, flow.Identifier{})
+	require.Equal(t, cache.idsFP, flow.Identifier{})
+	require.Equal(t, cache.chansFP, flow.Identifier{})
 	require.Empty(t, cache.cachedFanout)
 
 	// underlying topology should be called only once.
 	mock.AssertExpectationsForObjects(t, top)
 }
 
-// TestCache_Update evaluates that if input to GenerateFanout changes, the cache is invalidated and updated.
-func TestCache_Update(t *testing.T) {
+// TestCache_InputChange_IDs evaluates that if identity list input to GenerateFanout changes,
+// the cache is invalidated and updated.
+func TestCache_InputChange_IDs(t *testing.T) {
 	// mocks underlying topology returning a fanout
 	top := &mocknetwork.Topology{}
-	log := zerolog.New(os.Stderr).Level(zerolog.DebugLevel)
 	ids := unittest.IdentityListFixture(100)
 	fanout := ids.Sample(10)
-	top.On("GenerateFanout", ids).Return(fanout, nil).Once()
+	channels := network.ChannelList{"channel1", "channel2"}
+	top.On("GenerateFanout", mock.Anything, channels).Return(fanout, nil).Once()
 
-	// assumes cache holding some fanout
-	cache := NewCache(log, top)
+	// assumes cache holding some fanout for the same
+	// ids and channels defined above.
+	cache := NewCache(zerolog.Nop(), top)
 	cache.cachedFanout = ids.Sample(10)
-	cache.fingerprint = unittest.IdentifierFixture()
+	cache.idsFP = ids.Fingerprint()
+	cache.chansFP = channels.ID()
 
-	// cache content should change once fingerprint to GenerateFanout changes.
-	newFanout, err := cache.GenerateFanout(ids)
+	// cache content should change once input ids list to GenerateFanout changes.
+	// drops last id in the list to imitate a change.
+	newFanout, err := cache.GenerateFanout(ids[:len(ids)-1], channels)
 	require.NoError(t, err)
-	require.Equal(t, cache.fingerprint, ids.Fingerprint())
+	require.Equal(t, cache.idsFP, ids[:len(ids)-1].Fingerprint())
+	// channels input did not change, hence channels fingerprint should not be changed.
+	require.Equal(t, cache.chansFP, channels.ID())
+	// returned (new) fanout should be equal to the mocked fanout, and
+	// also should be cached.
+	require.Equal(t, cache.cachedFanout, fanout)
+	require.Equal(t, cache.cachedFanout, newFanout)
+
+	// underlying topology should be called only once.
+	mock.AssertExpectationsForObjects(t, top)
+}
+
+// TestCache_InputChange_Channels evaluates that if channel list input to GenerateFanout changes,
+// the cache is invalidated and updated.
+func TestCache_InputChange_Channels(t *testing.T) {
+	// mocks underlying topology returning a fanout
+	top := &mocknetwork.Topology{}
+	ids := unittest.IdentityListFixture(100)
+	fanout := ids.Sample(10)
+	channels := network.ChannelList{"channel1", "channel2"}
+	top.On("GenerateFanout", ids, mock.Anything).Return(fanout, nil).Once()
+
+	// assumes cache holding some fanout for the same
+	// ids and channels defined above.
+	cache := NewCache(zerolog.Nop(), top)
+	cache.cachedFanout = ids.Sample(10)
```

Comment:
same - I would rather make a call to `GenerateFanout` here instead of exploiting the inner working of the cache.